File: py/orbit_sdk/features.py
# Orbit SDK feature factory

from orbit_sdk.feature.base_feature import OrbitBaseFeature
from orbit_sdk.feature.debug_feature import OrbitDebugFeature
from orbit_sdk.feature.idempotency_feature import OrbitIdempotencyFeature
from orbit_sdk.feature.metrics_feature import OrbitMetricsFeature
from orbit_sdk.feature.paging_feature import OrbitPagingFeature
from orbit_sdk.feature.ratelimit_feature import OrbitRatelimitFeature
from orbit_sdk.feature.retry_feature import OrbitRetryFeature
from orbit_sdk.feature.test_feature import OrbitTestFeature
from orbit_sdk.feature.timeout_feature import OrbitTimeoutFeature


_FEATURES = {
    "base": lambda: OrbitBaseFeature(),
    "debug": lambda: OrbitDebugFeature(),
    "idempotency": lambda: OrbitIdempotencyFeature(),
    "metrics": lambda: OrbitMetricsFeature(),
    "paging": lambda: OrbitPagingFeature(),
    "ratelimit": lambda: OrbitRatelimitFeature(),
    "retry": lambda: OrbitRetryFeature(),
    "test": lambda: OrbitTestFeature(),
    "timeout": lambda: OrbitTimeoutFeature(),
}


def _make_feature(name):
    factory = _FEATURES.get(name)
    if factory is not None:
        return factory()
    return _FEATURES["base"]()


# True when this SDK was generated with the named feature class - the
# constructor's tolerance for extend-carried features reads this (an
# active name with no generated class must not become a BaseFeature
# stray when an extend instance carries it).
def _has_feature(name):
    return name in _FEATURES

File: php/features.php
<?php
declare(strict_types=1);

// Orbit SDK feature factory

require_once __DIR__ . '/feature/BaseFeature.php';
require_once __DIR__ . '/feature/DebugFeature.php';
require_once __DIR__ . '/feature/IdempotencyFeature.php';
require_once __DIR__ . '/feature/MetricsFeature.php';
require_once __DIR__ . '/feature/PagingFeature.php';
require_once __DIR__ . '/feature/RatelimitFeature.php';
require_once __DIR__ . '/feature/RetryFeature.php';
require_once __DIR__ . '/feature/TestFeature.php';
require_once __DIR__ . '/feature/TimeoutFeature.php';


class OrbitFeatures
{
    public static function make_feature(string $name)
    {
        switch ($name) {
            case "base":
                return new OrbitBaseFeature();
            case "debug":
                return new OrbitDebugFeature();
            case "idempotency":
                return new OrbitIdempotencyFeature();
            case "metrics":
                return new OrbitMetricsFeature();
            case "paging":
                return new OrbitPagingFeature();
            case "ratelimit":
                return new OrbitRatelimitFeature();
            case "retry":
                return new OrbitRetryFeature();
            case "test":
                return new OrbitTestFeature();
            case "timeout":
                return new OrbitTimeoutFeature();
            default:
                return new OrbitBaseFeature();
        }
    }

    /**
     * Does a generated feature class back this name? False for a name only
     * an options extend instance can supply (the station adopt path) - the
     * constructor uses this to skip make_feature for such names instead of
     * adding a stray BaseFeature.
     */
    public static function has_feature(string $name): bool
    {
        switch ($name) {
            case "base":
            case "debug":
            case "idempotency":
            case "metrics":
            case "paging":
            case "ratelimit":
            case "retry":
            case "test":
            case "timeout":
                return true;
            default:
                return false;
        }
    }
}
